R/brexit-map.R

```diff
@@ -1,56 +1,85 @@
 # Aim: create regional map of election results
+library(purrr)
 
 # Packages we'll use
 pkgs = c("rvest", "geojsonio", "tmap", "dplyr",
          "stringdist")
-lapply(pkgs, library, character.only = T)
+
+# @hrbrmstr note
+# walk() has no output side effects, so it's like a true list version
+# of a for loop
+walk(pkgs, library, character.only = T)
 
 # 1: Get local authority data
-url = "https://github.com/npct/pct-bigdata/raw/master/las-dbands.geojson"
-download.file(url, "las.geojson")
-las = geojson_read("las.geojson", what = "sp")
+
+# @hrbrmstr note
+# with url() being a function, many new folks to R have trouble
+# debugging errors when they've mistakenly used it in the wrong
+# context, so this helps provide a better practice for them.
+# also adding the "if" for the download as it's both bandwidth
+# friendly for the server and helps those R stats folks who aren't
+# on speedy connections (like some entire countries and even half of the
+# U.S. state I live in)
+URL = "https://github.com/npct/pct-bigdata/raw/master/las-dbands.geojson"
+fil = "las.geojson"
+if (!file.exists(fil)) download.file(URL, fil)
+las = geojson_read(fil, what = "sp")
 
 # 2: Get and test results data
-url = "http://www.bbc.co.uk/news/politics/eu_referendum/results/local/a"
-bbc = read_html(url)
+URL = "http://www.bbc.co.uk/news/politics/eu_referendum/results/local/a"
+bbc = read_html(URL)
 
 la_name = bbc %>%
-  html_nodes(".eu-ref-result-bar__title") %>% 
+  html_nodes(".eu-ref-result-bar__title") %>%
   html_text()
 vote_leave = bbc %>%
-  html_nodes(".eu-ref-result-bar__party-votes--percentage") %>% 
-  html_text() %>% 
-  gsub(pattern = " |%", replacement = "") %>% 
-  as.numeric() %>% 
+  html_nodes(".eu-ref-result-bar__party-votes--percentage") %>%
+  html_text() %>%
+  gsub(pattern = " |%", replacement = "") %>%
+  as.numeric()
 vote_leave = vote_leave[(1:length(la_name)) * 2 - 1]
 
 ref_result = data_frame(geo_label = la_name, `% vote leave` = vote_leave)
 
+# @hrbrmstr note: `local_authority` is not defined
+
 summary(local_authority %in% las$geo_label) # looks good
 las@data = left_join(las@data, ref_result)
 summary(las$vote_leave)
 qtm(las, "% vote leave")
 
 # 3: Run for all data points
-i = "b"
-for(i in letters[-1]){
-  url = "http://www.bbc.co.uk/news/politics/eu_referendum/results/local/"
-  url = paste0(url, i)
-  url_exists = RCurl::url.exists(url)
-  if(!url_exists) next
-  bbc = read_html(url)
+
+# @hrbrmstr note
+# avoids the need for RCurl URL check (see use inside map_df below)
+S_read_html <- safely(read_html)
+
+# @hrbrmstr note
+# this doesn't take too long but progress bars are always nice
+p <- progress_estimated(length(letters))
+
+# @hrbrmstr note
+# map_df() will let you return data[_.]frames and it will
+# efficiently do the rbinding. Also, since the BBC shows the
+# letters j, q, x & z grayed out, we can just not use them which
+# shld avoid any errant URL grabs.
+map_df(setdiff(letters, c("j", "q", "x", "z")), function(i) {
```

Comment:
Ah, it would def work, but since we actually know what's not there, might as well use that info (it's just a matter of a diff perspective).